osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -247,7 +254,71 @@ private EncryptedTypedSecret retrieveSecret(
     return Optional.of(secretsList.iterator().next());
   }
 
-  public synchronized void storeSecrets(
+  public synchronized void storeOrResetNewSecrets(
+      final String deviceIdentification, final List<TypedSecret> secrets) {
+    final List<TypedSecret> typedNewSecretsToStore = new ArrayList<>();
+    final List<TypedSecret> typedNewSecretsToReset = new ArrayList<>();
+    for (final TypedSecret typedSecret : secrets) {
+      final int numberOfNewSecrets =
+          this.secretRepository.getSecretCount(
+              deviceIdentification, typedSecret.getSecretType(), SecretStatus.NEW);
+      if (numberOfNewSecrets == 0) {
+        typedNewSecretsToStore.add(typedSecret);
+      } else if (this.recentNewSecretsPresent(
+          deviceIdentification, this.maxMinutesForNewKeyToBeActivated, typedSecret)) {
+        final String errorMsg =
+            "There is/are secrets of type %s for device %s with status NEW "
+                + "created less than %d minutes ago. No key with status NEW will be stored. Wait "
+                + "at least %d minutes before starting a request requiring NEW keys to be stored.";
+        throw new IllegalStateException(
+            String.format(
+                errorMsg,
+                typedSecret.getSecretType().name(),
+                deviceIdentification,
+                this.maxMinutesForNewKeyToBeActivated,
+                this.maxMinutesForNewKeyToBeActivated));
+      } else {
+        typedNewSecretsToReset.add(typedSecret);
+      }
+    }
+    this.storeSecrets(deviceIdentification, typedNewSecretsToStore);
+    typedNewSecretsToReset.forEach(
+        ts -> this.resetNewSecrets(deviceIdentification, ts.getSecretType()));
```

Comment:
We need a phonecall here

Reply:
Refactor: if multiple NEW keys then throw Exception(). Do NOT try to solve this on you one!!

Reply:
After discussion with @carlogeenen we decided that we can not think of a flow in which the situation of multiple NEW keys could emerge. So we should not try to solve this situation in this flow. We just take the most recent NEW key to continue the flow and leave other NEW keys untouched. We do not log this occurence and we do not throw an exception. This makes the replace key flow robust and successive replace key flows will use up all NEW keys in the end.

Reply:
In an earlier discussion I indicated that as far as I can tell there is no way to decide which key to use and which to expire as new keys in the database may or may not have been sent to the smart meter. You need to know which key is on the smart meter to be able to be able to set up the connection.

Because you can't choose and be certain you make the right choice, it is better to throw an exception.
If you can't think of a flow in which multiple keys can be in the database, the exception won't harm anyway.
If multiple new keys should not be allowed it may even be a good idea to enforce this with a database constraint (would need to be verified). This too would lead to exceptions in flows that should not occur, trying to introduce multiple new keys.

Reply:
In the beginning of the Replace Key Process an connection to the device is setup. If this connection fails on a key issue the Recover Key Process will be kicked off before finishing the Replace Key Process. This way we can be certain that a NEW key in SecretManagement is not on the device. And therefor we can not pick the wrong NEW key for replacement.